hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/OmDirectoryInfo.java

```diff
@@ -266,4 +267,26 @@ public OmDirectoryInfo copyObject() {
 
     return builder.build();
   }
+
+  /**
+   * Return a new copy of the object.
+   */
+  @Override
+  public Object clone() throws CloneNotSupportedException {
```

Comment:
I still have few questions here
1-CopyObject is implemented extensively in Ozone infor multiple classes.Seems the practice is wherever deep copy of object is required CopyObject is called.
Here where the issue is happening i.e in Liststatus method instead of making this change call CopyObject . For OmkeyInfo itself I can see couple of usages of CopyObject.
2- Also in case still we to implement this , is Clone way better or using copy constructor is better which removes overhead of implementing clonable and typecasting .

Reply:
> Here we are reading from cache which is of generic type and for this reason we cannot use CopyObject. ...

@ashishkumar50 , Could you show me the code?  I really think that we should not have both `copyObject()` and `clone()`.  Let me see if there is a way to get ride of `clone()`.

Reply:
We can use copy object at
https://github.com/apache/ozone/blob/7d1aaef989f3749014c870000c84a6645d6fb43b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java#LL317C1-L317C37
as keyInfo = ((OmKeyInfo) value).copyObject();
also for the directory OmDirectoryInfo dirInfo = ((OmDirectoryInfo) value).copyObject().
This can work for now.
But still in this class it will have referenced cache. And whenever someone use that cache directly in future without copyObject, It will lead to again corruption.
So instead just changing at this place to copyObject we decided to clone the object at the place where we are reading. So it will avoid any corruption in future.
Cache value is read at
https://github.com/apache/ozone/blob/7d1aaef989f3749014c870000c84a6645d6fb43b/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneListStatusHelper.java#L445

Reply:
Thanks for the info.  I would suggest that add a `CopyObject` interface (to define the copyObject() method) and don't use `ObjectUtils`.  Then, we can get ride of the `clone()` method.  (Alternatively, we may try to get ride of the `copyObject()`.  But it seems harder.)

Anyway, Ozone should not have both `clone()` and `copyObject()` methods at the same time.  It does not seem a good solution.

Reply:
Thanks for the suggestion, I will change to make use of CopyObject by defining new interface and remove clone.